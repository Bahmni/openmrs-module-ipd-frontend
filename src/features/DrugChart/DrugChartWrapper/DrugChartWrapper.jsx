import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import DrugChart from "../DrugChart";
import { useFetchMedications } from "../../../hooks/useFetchMedications";
export const TransformDrugChartData = (drugChartData) => {
  const drugOrderData = [];
  const slotDataByOrder = [];

  drugChartData.map((schedule) => {
    const { slots } = schedule;

    slots.forEach((slot) => {
      const slotData = {};
      const { startTime, status, order } = slot;
      const drugOrder = {
        uuid: order.uuid,
        drugName: order.drug.display,
        drugRoute: order.route.display,
        administrationInfo: [],
      };

      if (order.duration) {
        drugOrder.duration = order.duration + " " + order.durationUnits.display;
      }
      if (order.doseUnits.display !== "ml") {
        drugOrder.dosage = order.dose;
        drugOrder.doseType = order.doseUnits.display;
      } else {
        drugOrder.dosage = order.dose + order.doseUnits.display;
      }
      if (order.duration) {
        drugOrder.duration = order.duration + " " + order.durationUnits.display;
      }

      const startDateTimeObj = new Date(startTime * 1000);
      let adminInfo, administeredTime;
      if (slot.admin) {
        const { administeredBy, administeredAt } = slot.admin;
        administeredTime = moment(new Date(administeredAt)).format("HH:mm");
        adminInfo = administeredBy + " [" + administeredTime + "]";
      } else {
        adminInfo = "";
      }
      const startHour = startDateTimeObj.getHours();
      const startMinutes = startDateTimeObj.getMinutes();
      slotData[startHour] = {
        minutes: startMinutes,
        status: status,
        administrationInfo: adminInfo,
      };
      if (status === "Administered" || status === "Administered-Late") {
        drugOrder.administrationInfo.push({
          kind: status,
          time: administeredTime,
        });
      }
      if (
        !drugOrderData.some(
          (existingOrder) =>
            existingOrder.drugName === drugOrder.drugName &&
            existingOrder.uuid === drugOrder.uuid
        )
      ) {
        drugOrderData.push(drugOrder);
        slotDataByOrder.push(slotData);
      } else {
        const index = drugOrderData.findIndex(
          (existingOrder) =>
            existingOrder.drugName === drugOrder.drugName &&
            existingOrder.uuid === drugOrder.uuid
        );
        slotDataByOrder[index] = {
          ...slotDataByOrder[index],
          ...slotData,
        };
      }
    });
  });
  return [slotDataByOrder, drugOrderData];
};
export const GetUTCEpochForDate = (viewDate) => {
  const utcTimeEpoch = moment.utc(viewDate).unix();
  return utcTimeEpoch;
};
export default function DrugChartWrapper(props) {
  const { patientId, viewDate } = props;
  const forDate = GetUTCEpochForDate(viewDate);

  const { drugChartData, isLoading } = useFetchMedications(patientId, forDate);
  const transformedDrugChartData = TransformDrugChartData(drugChartData);

  if (isLoading) {
    return <div>Loading...</div>;
  }
  return <DrugChart drugChartData={transformedDrugChartData} />;
}
DrugChartWrapper.propTypes = {
  patientId: PropTypes.string.isRequired,
  viewDate: PropTypes.instanceOf(Date),
};
