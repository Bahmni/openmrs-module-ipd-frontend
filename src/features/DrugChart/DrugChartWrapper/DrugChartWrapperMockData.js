export const drugChartData = [
  {
    id: 1,
    uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
    serviceType: "Medication Administration",
    comment: "some comment",
    startDate: 1690906304,
    endDate: 1691165503,
    patient: {
      uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
    },
    order: {
      drug: {
        display: "Paracetamol 120 mg/5 mL Suspension (Liquid)",
      },
      route: {
        uuid: "9d6bc13f-3f10-11e4-adec-0800271c1b75",
        display: "Oral",
      },
      dose: 25,
      doseUnits: {
        uuid: "86239663-7b04-4563-b877-d7efc4fe6c46",
        display: "ml",
      },
      duration: 3,
      durationUnits: {
        uuid: "9d7437a9-3f10-11e4-adec-0800271c1b75",
        display: "Day(s)",
      },
    },
    slots: [
      {
        id: 1,
        uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
        orderId: 11,
        serviceType: "MedicationRequest",
        status: "SCHEDULED",
        startTime: 1690906550,
      },
      {
        id: 2,
        uuid: "738aa77d-03fc-438f-a87a-ae8a8867cyu721",
        orderId: 11,
        serviceType: "MedicationRequest",
        status: "SCHEDULED",
        startTime: 1690916550,
      },
      {
        id: 3,
        uuid: "738aa77d-03fc-438f-a87a-a3ed8a8867c421",
        serviceType: "MedicationRequest",
        status: "SCHEDULED",
        startTime: 1690918550,
      },
    ],
  },
];
