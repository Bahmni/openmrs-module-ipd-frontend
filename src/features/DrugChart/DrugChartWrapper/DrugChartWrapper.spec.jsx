import React from "react";
import { render } from "@testing-library/react";
import DrugChartWrapper, { TransformDrugChartData } from "./DrugChartWrapper";

const drugChartData = [
  {
    scheduleId: 1,
    scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
    scheduleServiceType: "Medication Administration",
    patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
    comment: "some comment",
    startDate: "2023-08-08T18:30:00.000Z",
    endDate: "2023-08-08T18:30:00.000Z",
    order: {
      drug: {
        display: "Paracetamol 120 mg/5 mL Suspension (Liquid)",
      },
      route: {
        uuid: "9d6bc13f-3f10-11e4-adec-0800271c1b75",
        display: "Oral",
      },
      dose: 25,
      doseUnits: {
        uuid: "86239663-7b04-4563-b877-d7efc4fe6c46",
        display: "ml",
      },
      duration: 3,
      durationUnits: {
        uuid: "9d7437a9-3f10-11e4-adec-0800271c1b75",
        display: "Day(s)",
      },
    },
    slots: [
      {
        id: 1,
        uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
        orderId: 11,
        serviceType: "MedicationRequest",
        status: "SCHEDULED",
        startTime: 1690906550,
      },
      {
        id: 2,
        uuid: "738aa77d-03fc-438f-a87a-ae8a8867cyu721",
        orderId: 11,
        serviceType: "MedicationRequest",
        status: "SCHEDULED",
        startTime: 1690916550,
      },
      {
        id: 3,
        uuid: "738aa77d-03fc-438f-a87a-a3ed8a8867c421",
        serviceType: "MedicationRequest",
        status: "SCHEDULED",
        startTime: 1690918550,
      },
    ],
  },
];
describe("DrugChartWrapper", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <DrugChartWrapper drugChartData={drugChartData} />
    );
    expect(container).toMatchSnapshot();
  });
});

describe("TransformDrugChartData", () => {
  it("should transform drug chart data", () => {
    const TransformedDrugChartData = TransformDrugChartData(drugChartData);
    expect(TransformedDrugChartData).toEqual([
      [
        {
          16: {
            administrationInfo: "",
            minutes: 15,
            status: "SCHEDULED",
          },
          19: {
            administrationInfo: "",
            minutes: 35,
            status: "SCHEDULED",
          },
        },
      ],
      [
        {
          administrationInfo: [],
          dosage: "25ml",
          drugName: "Paracetamol 120 mg/5 mL Suspension (Liquid)",
          drugRoute: "Oral",
          duration: "3 Day(s)",
        },
      ],
    ]);
  });
});
