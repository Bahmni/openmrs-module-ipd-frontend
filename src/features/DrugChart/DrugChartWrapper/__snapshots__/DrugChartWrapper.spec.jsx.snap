// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugChartWrapper matches snapshot 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
